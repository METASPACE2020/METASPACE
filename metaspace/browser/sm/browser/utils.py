import pathlib
import shutil


def list_file_sizes(bucket):
    obj_sizes = {}
    i = 0
    for obj in bucket.objects.all():
        obj_sizes[obj.key] = obj.size
        i += 1
        if i % 1000:
            print(f"{i} objects")

    obj_sizes = sorted(obj_sizes.items(), key=lambda kv: kv[1], reverse=True)

    obj_sizes_5gbmax = [(key, size) for key, size in obj_sizes if size < 5 * 1024 ** 3]
    for i in range(10):
        key, size = obj_sizes_5gbmax[i]
        print(size / 1024 ** 2, key)


def download_dataset(bucket, key_prefix, ds_local_path):
    ds_local_path.mkdir(parents=True, exist_ok=True)
    for obj in bucket.objects.filter(Prefix=key_prefix):
        # print(obj)
        fn = obj.key.split('/')[-1]
        f_path = ds_local_path / fn
        if not f_path.exists():
            bucket.download_file(obj.key, str(f_path))


def get_file_by_ext(path, ext):
    return [f_path for f_path in path.iterdir() if str(f_path).lower().endswith(ext)][0]


def clean_dir(path):
    shutil.rmtree(path, ignore_errors=True)
    path.mkdir(parents=True, exist_ok=True)


def mz_ppm_bin(mz, ppm):
    return mz - mz * ppm * 1e-6, mz + mz * ppm * 1e-6
