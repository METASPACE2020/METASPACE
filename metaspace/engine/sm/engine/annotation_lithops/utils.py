from __future__ import annotations

import json
import logging
from base64 import urlsafe_b64encode
from datetime import datetime
from hashlib import blake2b
from pathlib import Path

import numpy as np
import pandas as pd

# logging.getLogger('ibm_boto3').setLevel(logging.WARNING)
# logging.getLogger('ibm_botocore').setLevel(logging.WARNING)
# logging.getLogger('urllib3').setLevel(logging.WARNING)

logger = logging.getLogger('annotation-pipeline')


def ds_dims(coordinates):
    min_x, min_y = np.amin(coordinates, axis=0)[:2]
    max_x, max_y = np.amax(coordinates, axis=0)[:2]
    nrows, ncols = max_y - min_y + 1, max_x - min_x + 1
    return nrows, ncols


def get_pixel_indices(coordinates):
    _coord = np.array(coordinates)[:, :2]
    _coord = np.around(_coord, 5)
    _coord -= np.amin(_coord, axis=0)

    _, ncols = ds_dims(coordinates)
    pixel_indices = _coord[:, 1] * ncols + _coord[:, 0]
    pixel_indices = pixel_indices.astype(np.int32)
    return pixel_indices


def jsonhash(obj) -> str:
    """
    Calculates a hash for a JSON-stringifiable object. Intended for compacting large sets of
    parameters into a simple key that can be used to distinctly identify a cache entry.

    The output is collision-resistant, but shouldn't be assumed to be cryptographically secure.
    In most cases a motivated adversary could figure out the original object contents easily, as
    there's no hidden key and it's unlikely there will be much variety in the objects hashed.
    """
    json_val = json.dumps(obj, sort_keys=True)
    hash_val = blake2b(json_val.encode(), digest_size=12).digest()
    return str(urlsafe_b64encode(hash_val), 'utf-8')
