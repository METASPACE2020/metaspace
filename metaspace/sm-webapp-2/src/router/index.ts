import { createRouter, createWebHistory } from 'vue-router'
import { Component } from 'vue'
import NotFound from '../modules/App/NotFoundPage.vue'
import AboutPage from '../modules/App/AboutPage'
import DatasetsPage from '../modules/Datasets/DatasetsPage.vue'

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ '../modules/Annotations/AnnotationsPage.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ '../modules/MetadataEditor/MetadataEditPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ '../modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/imzml/DatasetBrowserPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/HelpPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Project/ProjectsListPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/PrivacyPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Project/ViewProjectPage.vue'),

}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, Component>

export const routes : any =[
  { path: '/', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/about', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
  {
    path: '/datasets',
    component: DatasetsPage,
    children: [
      { path: '', component: asyncPages.DatasetTable },
      { path: 'summary', component: asyncPages.DatasetTable },
    ],
  },
  {
    path: '/datasets/:dataset_id/comparison',
    name: 'datasets-comparison',
    component: asyncPages.DatasetComparisonPage,
  },
  { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
  { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.DatasetTable },
  { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
  { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
  { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
  { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetTable },
  { path: '/upload', component: asyncPages.UploadPage },
  { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },



  { path: '/groups', component: asyncPages.GroupsListPage },
  { path: '/group/create', component: asyncPages.CreateGroupPage },
  { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },


  { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
  { // Legacy URL sent in "request access" emails up until Feb 2019
    path: '/project/:projectIdOrSlug/manage',
    redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } },
  },
  { path: '/projects', component: asyncPages.ProjectsListPage },



  { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },


  {
    path: '/:pathMatch(.*)*',
    component: NotFound,
  },
]

const router = createRouter({
  // @ts-ignore
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
  scrollBehavior(to, from, savedPosition) {
    // return desired position
    if (savedPosition) {
      return savedPosition;
    }
    return { top: 0 };
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})


const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
