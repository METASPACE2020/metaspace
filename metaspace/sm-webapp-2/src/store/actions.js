import apolloClient from '../api/graphqlClient';
import {fetchOptionListsQuery} from '../api/metadata';
import {decodeParams} from '../modules/Filters';
import {computed} from "vue";
import store from "../store/index";
import reportError from "@/lib/reportError";
export default {

  async initFilterLists(context) {
    if(context.state.filterListsLoading || context.state.filterLists != null)
      return;

    context.commit('setFilterListsLoading');
    let response = {}

    try{
      response = await apolloClient.query({
        query: fetchOptionListsQuery,
        fetchPolicy: 'cache-first',
      });
    } catch (err) {
      reportError(err);
    }

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds)} );

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)
    context.commit('updateFilter', filter)
  },

};
