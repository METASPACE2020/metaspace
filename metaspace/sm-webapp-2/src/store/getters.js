export default {
  filter() {
    return {}
    // return decodeParams(state.route, state.filterLists);
  },

  currentUser(state) {
    return state.currentUser;
  },
}
