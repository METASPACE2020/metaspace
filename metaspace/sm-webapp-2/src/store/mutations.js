import { isEqual, omit, without } from 'lodash-es'
import router from '../router'
import compare from '../lib/compare'
import {
  decodeParams,
  encodeParams,
  encodeSections,
  encodeSortOrder,
  FILTER_SPECIFICATIONS,
  getFilterInitialValue,
  stripFilteringParams,
} from '../modules/Filters'
import { DEFAULT_ANNOTATION_VIEW_SECTIONS, DEFAULT_COLORMAP, DEFAULT_TABLE_ORDER } from '../modules/Filters/url'
import { DEFAULT_SCALE_TYPE } from '../lib/constants'

function updatedLocation(state, filter) {
  const query = encodeParams(filter, state.route.path, state.filterLists)

  state.lastUsedFilters[state.route.path] = {
    filter,
    query,
    order: state.orderedActiveFilters,
  }

  return {
    query: Object.assign(query, stripFilteringParams(state.route.query)),
  }
}

function routerReplace(state, params) {
  router.replace(params)
  state.route.path = params.path || state.route.path
  state.route.params = params.params || state.route.params
  state.route.query = params.query || state.route.query
}

function routerPush(state, params) {
  router.push(params)
  state.route.path = params.path || state.route.path
  state.route.params = params.params || state.route.params
  state.route.query = params.query || state.route.query
}

function replaceURL(state, filter) {
  routerReplace(state, updatedLocation(state, filter))
}

function pushURL(state, filter) {
  routerPush(state, updatedLocation(state, filter))
}

function sortFilterKeys(keys) {
  keys.sort((a, b) => compare(FILTER_SPECIFICATIONS[a].sortOrder || 100, FILTER_SPECIFICATIONS[b].sortOrder || 100))
}

function updateFilter(state, filter, routerAction = null) {
  const active = []

  // drop unset filters
  for (let i = 0; i < state.orderedActiveFilters.length; i++) {
    const key = state.orderedActiveFilters[i]
    if (filter[key] !== undefined) active.push(key)
  }

  // append newly added filters to the end
  for (const key in filter) if (filter[key] !== undefined && active.indexOf(key) == -1) active.push(key)

  // sort
  sortFilterKeys(active)

  const changedFilterSet = !isEqual(state.orderedActiveFilters, active)

  state.orderedActiveFilters = active
  if (routerAction === 'push' || (routerAction == null && changedFilterSet)) pushURL(state, filter)
  else replaceURL(state, filter)
}

export default {
  updateRoute(state, route) {
    state.route.path = route.path
    state.route.params = route.params
    state.route.query = route.query
  },

  updateFilter(state, filter) {
    // TODO: Stop using updateFilter for adding/removing filters, as it makes reacting to added/removed filters
    // much more difficult
    updateFilter(state, filter)
  },

  updateCurrentUser(state, user) {
    state.currentUser = user
  },

  setReportError(state) {
    state.reportError = !state.reportError
  },

  replaceFilter(state, filter) {
    updateFilter(state, filter, 'replace')
  },

  addFilter(state, name) {
    const oldFilter = decodeParams(state.route, state.filterLists)
    const filtersToAdd = [name]
    const filtersToRemove = []

    // Check for required additional filters & conflicting filters
    ;(FILTER_SPECIFICATIONS[name].dependsOnFilters || []).forEach((key) => {
      if (oldFilter[key] === undefined) {
        filtersToAdd.push(key)
      }
    })
    ;(FILTER_SPECIFICATIONS[name].conflictsWithFilters || []).forEach((key) => {
      if (oldFilter[key] !== undefined) {
        filtersToRemove.push(key)
      }
    })

    const newFilter = { ...oldFilter }
    filtersToAdd.forEach((key) => {
      newFilter[key] = getFilterInitialValue(key, state.filterLists)
    })
    filtersToRemove.forEach((key) => {
      newFilter[key] = undefined
    })
    const newActive = without([...state.orderedActiveFilters, ...filtersToAdd], ...filtersToRemove)
    sortFilterKeys(newActive)

    state.orderedActiveFilters = newActive
    pushURL(state, newFilter)
  },

  removeFilter(state, name) {
    const oldFilter = decodeParams(state.route, state.filterLists)
    const filtersToRemove = [name]

    // Check for dependent filters that should also be removed
    Object.keys(oldFilter).forEach((key) => {
      const { dependsOnFilters } = FILTER_SPECIFICATIONS[key]
      if (dependsOnFilters != null && dependsOnFilters.includes(name)) {
        filtersToRemove.push(key)
      }
    })

    const newFilter = omit(oldFilter, filtersToRemove)
    state.orderedActiveFilters = without(state.orderedActiveFilters, ...filtersToRemove)
    pushURL(state, newFilter)
  },

  updateFilterOnNavigate(state, to) {
    const newActiveFilters = Object.keys(decodeParams(to, state.filterLists))

    const removed = without(state.orderedActiveFilters, ...newActiveFilters)
    const added = without(newActiveFilters, ...state.orderedActiveFilters)

    state.orderedActiveFilters = without(state.orderedActiveFilters, ...removed)
    state.orderedActiveFilters.push(...added)
    sortFilterKeys(state.orderedActiveFilters)
  },

  setFilterListsLoading(state) {
    state.filterListsLoading = true
  },

  setFilterLists(state, filterLists) {
    state.filterLists = filterLists
    state.filterListsLoading = false
  },

  setAnnotation(state, annotation) {
    state.annotation = annotation
  },

  setNormalizationMatrix(state, normalizationMatrix) {
    state.normalization = normalizationMatrix
  },

  resetChannels(state) {
    state.channels = []
  },

  addChannel(state, { id, annotations, settings }) {
    state.channels.push({ id, annotations, settings })
  },

  restoreChannels(state, channels) {
    state.channels = channels
  },

  removeChannel(state, { index }) {
    const aux = state.channels.slice(0)
    aux.splice(index, 1)
    state.channels = aux
  },

  updateChannel(state, { index, id, annotations, settings }) {
    const aux = state.channels.slice(0)
    aux[index] = { id, annotations, settings }
    state.channels = aux
  },

  setViewerMode(state, mode = 'SINGLE') {
    state.mode = mode
  },

  setRoiInfo(state, { key, roi }) {
    state.roiInfo[key] = roi
  },

  toggleRoiVisibility(state, visible) {
    state.roiInfo['visible'] = visible
  },

  resetRoiInfo(state) {
    state.roiInfo = { visible: false }
  },

  setSnapshotAnnotationIds(state, annotation) {
    state.snapshotAnnotationIds = annotation
  },

  updateAnnotationTableStatus(state, isLoading) {
    state.tableIsLoading = isLoading
  },

  startTour(state, tourDescription) {
    state.currentTour = tourDescription
  },

  endTour(state) {
    state.currentTour = null
  },

  updateAnnotationViewSections(state, activeSections) {
    const sections = encodeSections(activeSections)
    const defaultSections = encodeSections(DEFAULT_ANNOTATION_VIEW_SECTIONS)
    routerReplace(state, {
      query: sections !== defaultSections ? { ...state.route.query, sections } : omit(state.route.query, 'sections'),
    })
  },

  setColormap(state, cmap) {
    routerReplace(state, {
      query: cmap !== DEFAULT_COLORMAP ? { ...state.route.query, cmap } : omit(state.route.query, 'cmap'),
    })
  },

  setScaleType(state, scaleType) {
    routerReplace(state, {
      query:
        scaleType !== DEFAULT_SCALE_TYPE
          ? { ...state.route.query, scale: scaleType }
          : omit(state.route.query, 'scale'),
    })
  },

  setLockTemplate(state, lockTemplate) {
    routerReplace(state, {
      query: lockTemplate ? { ...state.route.query, lock: lockTemplate } : omit(state.route.query, 'lock'),
    })
  },

  setNormalization(state, normalization) {
    routerReplace(state, {
      query: !normalization ? omit(state.route.query, 'norm') : { ...state.route.query, norm: normalization },
    })
  },

  setCurrentPage(state, page) {
    routerReplace(state, {
      query: page !== 1 ? { ...state.route.query, page: String(page) } : omit(state.route.query, 'page'),
    })
  },

  setRow(state, row) {
    routerReplace(state, {
      query: row !== 1 ? { ...state.route.query, row: String(row) } : omit(state.route.query, 'row'),
    })
  },

  setSortOrder(state, sortOrder) {
    const sort = encodeSortOrder(sortOrder)
    const defaultSort = encodeSortOrder(DEFAULT_TABLE_ORDER)
    routerReplace(state, {
      query: sort !== defaultSort ? { ...state.route.query, sort } : omit(state.route.query, 'sort'),
    })
  },

  setColocalizationAlgo(state, colocalizationAlgo) {
    routerReplace(state, {
      query:
        colocalizationAlgo != null ? { ...state.route.query, alg: colocalizationAlgo } : omit(state.route.query, 'alg'),
    })
  },

  setDatasetTab(state, tab) {
    const path = {
      list: '/datasets',
      summary: '/datasets/summary',
    }[tab.toLowerCase()]
    routerReplace(state, {
      path,
      query: state.route.query,
    })
  },
}
