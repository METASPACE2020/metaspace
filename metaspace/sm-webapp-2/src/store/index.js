import { createStore } from 'vuex'
import mutations from './mutations'
import getters from './getters'
import actions from './actions.js'

import { accountModule } from '../modules/Account'

const store = createStore({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    reportError: false,

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: { visible: false },

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE',
    route: {
      path: '',
      params: {},
      query: {},
    },
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule,
  },
})

export default store
