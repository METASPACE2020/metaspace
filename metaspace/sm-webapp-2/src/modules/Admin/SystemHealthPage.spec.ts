import { nextTick, ref } from 'vue'
import { flushPromises, mount } from '@vue/test-utils'
import SystemHealthPage from './SystemHealthPage.vue'
import { initMockGraphqlClient } from '../../tests/utils/mockGraphqlClient'
import { vi } from 'vitest'
import { DefaultApolloClient, useMutation, useQuery } from '@vue/apollo-composable'
import store from '../../store'
import router from '../../router'

vi.mock('@vue/apollo-composable', () => ({
  useQuery: vi.fn(),
  useMutation: vi.fn(),
  useSubscription: vi.fn(() => ({ onResult: vi.fn() })),
  DefaultApolloClient: vi.fn(),
}))

let graphqlMocks: any

describe('SystemHealthPage', () => {
  const mockGraphql = async (qyeryParams, mutationParams) => {
    graphqlMocks = await initMockGraphqlClient({
      Query: () => qyeryParams,
      Mutation: () => mutationParams,
    })
    ;(useQuery as any).mockReturnValue({
      result: ref(Object.keys(qyeryParams).reduce((acc, key) => ({ ...acc, [key]: qyeryParams[key]() }), {})),
      loading: ref(false),
      onResult: vi.fn(),
      subscribeToMore: vi.fn(),
    })
    ;(useMutation as any).mockReturnValue({
      mutate: mutationParams,
    })
  }
  const adminQuery = async () => {
    const queryParams = {
      currentUser: () => ({ id: 'userid', role: 'admin' }),
    }
    const mutationParams = {
      updateSystemHealth: () => true,
    }
    await mockGraphql(queryParams, mutationParams)
  }
  const updateErrorQuery = async () => {
    const queryParams = {
      currentUser: () => ({ id: 'userid', role: 'admin' }),
    }
    const mutationParams = {
      updateSystemHealth: () => new Error('internal error'),
    }
    await mockGraphql(queryParams, mutationParams)
  }
  const userQuery = async () => {
    const queryParams = {
      currentUser: () => ({ id: 'userid', role: 'user' }),
    }
    await mockGraphql(queryParams, {})
  }

  it('should match snapshot logged as non-admin', async () => {
    await userQuery()
    const wrapper = mount(SystemHealthPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should match snapshot logged as admin', async () => {
    await adminQuery()
    const wrapper = mount(SystemHealthPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to toggle all system health options as admin', async () => {
    await adminQuery()
    const wrapper = mount(SystemHealthPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    wrapper.vm.mode = 0
    await nextTick()
    expect(wrapper.findAll('.el-radio').at(0).classes()).toContain('is-checked')

    wrapper.vm.mode = 1
    await nextTick()
    expect(wrapper.findAll('.el-radio').at(1).classes()).toContain('is-checked')

    wrapper.vm.mode = 2
    await nextTick()
    expect(wrapper.findAll('.el-radio').at(2).classes()).toContain('is-checked')
  })

  it('should toggle Enable everything option and update', async () => {
    await adminQuery()
    const wrapper = mount(SystemHealthPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    wrapper.vm.mode = 0
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Update')
    wrapper.find('.el-button').trigger('click')
    await flushPromises()
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Updated')
  })

  it('should toggle Disable dataset upload/reprocessing option and update', async () => {
    await adminQuery()
    const wrapper = mount(SystemHealthPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    wrapper.vm.mode = 1
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Update')
    wrapper.find('.el-button').trigger('click')
    await flushPromises()
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Updated')
  })

  it('should toggle Read-only mode option and update', async () => {
    await adminQuery()
    const wrapper = mount(SystemHealthPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    wrapper.vm.mode = 2
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Update')
    wrapper.find('.el-button').trigger('click')
    await flushPromises()
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Updated')
  })

  it('should display update error', async () => {
    await updateErrorQuery()
    const wrapper = mount(SystemHealthPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    wrapper.vm.mode = 2
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Update')
    wrapper.find('.el-button').trigger('click')
    await flushPromises()
    await nextTick()

    expect(wrapper.find('.el-button').text()).toBe('Error')
  })
})
