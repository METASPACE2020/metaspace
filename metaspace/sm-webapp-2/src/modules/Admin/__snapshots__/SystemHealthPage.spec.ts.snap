// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SystemHealthPage > should match snapshot logged as admin 1`] = `
"<div data-v-979003de=\\"\\" class=\\"mx-auto max-w-4xl\\">
  <form data-v-979003de=\\"\\" class=\\"el-form el-form--default el-form--label-top\\">
    <div data-v-979003de=\\"\\" class=\\"el-form-item asterisk-left\\">
      <!--v-if-->
      <div class=\\"el-form-item__content\\">
        <div data-v-979003de=\\"\\" class=\\"el-radio-group radio-group-wrapper\\" role=\\"radiogroup\\" aria-labelledby=\\"el-id-replaced\\" id=\\"el-id-replaced\\"><label data-v-979003de=\\"\\" class=\\"el-radio is-bordered is-checked\\"><span class=\\"el-radio__input is-checked\\"><input class=\\"el-radio__original\\" name=\\"el-id-replaced\\" type=\\"radio\\" value=\\"0\\"><span class=\\"el-radio__inner\\"></span></span><span class=\\"el-radio__label\\"> Enable everything </span></label><label data-v-979003de=\\"\\" class=\\"el-radio is-bordered\\"><span class=\\"el-radio__input\\"><input class=\\"el-radio__original\\" name=\\"el-id-replaced\\" type=\\"radio\\" value=\\"1\\"><span class=\\"el-radio__inner\\"></span></span><span class=\\"el-radio__label\\"> Disable dataset upload/reprocessing </span></label><label data-v-979003de=\\"\\" class=\\"el-radio is-bordered\\"><span class=\\"el-radio__input\\"><input class=\\"el-radio__original\\" name=\\"el-id-replaced\\" type=\\"radio\\" value=\\"2\\"><span class=\\"el-radio__inner\\"></span></span><span class=\\"el-radio__label\\"> Read-only mode </span></label></div>
      </div>
    </div>
    <div data-v-979003de=\\"\\" class=\\"el-form-item asterisk-left\\"><label id=\\"el-id-replaced\\" for=\\"el-id-replaced\\" class=\\"el-form-item__label\\">Maintenance message (optional)</label>
      <div class=\\"el-form-item__content\\">
        <div data-v-979003de=\\"\\" class=\\"el-input\\" style=\\"width: 450px;\\">
          <!-- input -->
          <!-- prepend slot -->
          <!--v-if-->
          <div class=\\"el-input__wrapper\\" tabindex=\\"-1\\">
            <!-- prefix slot -->
            <!--v-if--><input class=\\"el-input__inner\\" type=\\"text\\" autocomplete=\\"off\\" tabindex=\\"0\\" id=\\"el-id-replaced\\"><!-- suffix slot -->
            <!--v-if-->
          </div><!-- append slot -->
          <!--v-if-->
        </div>
      </div>
    </div><button data-v-979003de=\\"\\" ariadisabled=\\"false\\" type=\\"button\\" class=\\"el-button el-button--primary\\">
      <!--v-if--><span class=\\"\\">Update</span>
    </button>
  </form>
</div>"
`;

exports[`SystemHealthPage > should match snapshot logged as non-admin 1`] = `"<!--v-if-->"`;
